### DISCUSSION TITLE
finances API

### DISCUSSION THREAD

**Discussion by zyzjian** [saleweaver/python-amazon-sp-api#159] [FAQ]
I'm getting data through the Finances API, but when I pass in the next_token page flip, it returns exactly the same data as the previous page.

(The different Next_Tokens return exactly the same data)
Did I get it wrong?

**Comment by zyzjian** [comment]
How do you get the financial data? How do you pass the parameters in? I'm a little confused about these.I hope someone can help me about this~ 
thanks!!!

**Reply by yberezkin** [reply]
May you post your request here? Will all parameters you use? Thanks.

**Reply by zyzjian** [reply]
> May you post your request here? Will all parameters you use? Thanks.

data = Finances(credentials=self.credentials).list_financial_events(
                PostedAfter=self.dateafter,
                PostedBefore=self.datebefore,
                next_token=next_token
            )
Financial_Events_By_GroupId = Finances(credentials=self.credentials).list_financial_events_by_group_id(
                    event_group_id=FinancialEventGroupId[0],
                    next_token=next_token
                )

**this is my request code. am i wrong? thanks for your help~~**

**Reply by yberezkin** [reply]
In this code you don't read, update the next_token variable.
In general, after you made the first request, if the response has NextToken - then you make a new call with the next_token variable. Moreover - the API argument name  in `Finances(nextToken)` not `Finances(next_token)`

**Reply by zyzjian** [reply]
> In this code you don't read, update the next_token variable.
> In general, after you made the first request, if the response has NextToken - then you make a new call with the next_token variable. Moreover - the API argument name in `Finances(nextToken)` not `Finances(next_token)`

sorry ~My last comment seems to have been swallowed up by the system.I thought I had submitted my reply successfully.
I tried it as you suggested,But it doesn't work.
this is my code
`next_token =None
while True:
       Financial_Events_By_GroupId = Finances(credentials=self.credentials).list_financial_events_by_group_id(
       event_group_id=FinancialEventGroupId,
       nextToken=next_token
                )
       next_token = Financial_Events_By_GroupId.next_token`

Can you show me some sample codes~

**Reply by zyzjian** [reply]
> In this code you don't read, update the next_token variable.
> In general, after you made the first request, if the response has NextToken - then you make a new call with the next_token variable. Moreover - the API argument name in `Finances(nextToken)` not `Finances(next_token)`

sorry ~My last comment seems to have been swallowed up by the system.I thought I had submitted my reply successfully.
I tried it as you suggested,But it doesn't work.
this is my code
`next_token =None
while True:
       Financial_Events_By_GroupId = Finances(credentials=self.credentials).list_financial_events_by_group_id(
       event_group_id=FinancialEventGroupId,
       nextToken=next_token
                )
       next_token = Financial_Events_By_GroupId.next_token`

Can you show me some sample codes~

**Reply by yberezkin** [reply]
```python3
date = datetime.now() - timedelta(days=7)
date = date.replace(hour=0, minute=0, second=0, microsecond=0).isoformat()
response = Finances(credentials=CREDS).list_financial_events(PostedAfter=date, MaxResultsPerPage=100)
next_token = True
results = []
count = 0
while next_token:
    fin_events = {each: response.payload['FinancialEvents'][each] for each in response.payload['FinancialEvents']
                  if response.payload['FinancialEvents'][each]}
    results.append(fin_events)
    next_token = response.next_token if response.next_token is not None else None
    if next_token:
        response = Finances(credentials=CREDS).list_financial_events(NextToken=next_token)
```
Just ran it today - June 30, 2021 over master branch. Also - the param seems to be `NextToken`. Works well, never tried actually by_Group_id, but believe it would be the same.

Here is -> [General AWS documentation](https://github.com/amzn/selling-partner-api-docs/blob/main/references/finances-api/financesV0.md#listfinancialeventsbygroupid)

**Reply by zyzjian** [reply]
> ```python
> date = datetime.now() - timedelta(days=7)
> date = date.replace(hour=0, minute=0, second=0, microsecond=0).isoformat()
> response = Finances(credentials=CREDS).list_financial_events(PostedAfter=date, MaxResultsPerPage=100)
> next_token = True
> results = []
> count = 0
> while next_token:
>     fin_events = {each: response.payload['FinancialEvents'][each] for each in response.payload['FinancialEvents']
>                   if response.payload['FinancialEvents'][each]}
>     results.append(fin_events)
>     next_token = response.next_token if response.next_token is not None else None
>     if next_token:
>         response = Finances(credentials=CREDS).list_financial_events(NextToken=next_token)
> ```
> 
> Just ran it today - June 30, 2021 over master branch. Also - the param seems to be `NextToken`. Works well, never tried actually by_Group_id, but believe it would be the same.
> 
> Here is -> [General AWS documentation](https://github.com/amzn/selling-partner-api-docs/blob/main/references/finances-api/financesV0.md#listfinancialeventsbygroupid)

thanks!!! i get it !
